src/libraries/Microsoft.PowerFx.Core/Public/Marshal/ITypeMashalerProvider.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using Microsoft.PowerFx.Core.Public.Types;
+using Microsoft.PowerFx.Core.Public.Values;
+
+namespace Microsoft.PowerFx
+{
+    /// <summary>
+    /// Handles marshaling a given type. Invoked by the <see cref="TypeMarshallerCache"/>.
+    /// </summary>
+    public interface ITypeMashalerProvider
+    {
+        // Return null if it doesn't handle it. 
+        // A single ITypeMarshaler can be created once per type and then reused for each instance.
+        // Pass in a cache for aggregate types that need to marshal sub types. 
+        public ITypeMarshaler New(Type type, TypeMarshallerCache cache, int maxDepth);
+    }
+
+    /// <summary>
+    /// A mrashaller for a given System.Type. 
+    /// </summary>
+    public interface ITypeMarshaler
+    {
+        public FormulaType Type { get; }
```

Comment:
>public FormulaType Type { get; } [](http://example.com/codeflow?start=8&length=32)

subtle win here - the Type is known regardless of the runtime object (which is only passed to Marshal).